src/java_bytecode/generic_parameter_specialization_map_keys.cpp

```diff
@@ -0,0 +1,152 @@
+{
+  PRECONDITION(
+    is_java_generic_class_type(type) ||
+    is_java_implicitly_generic_class_type(type));
+  std::vector<java_generic_parametert> generic_parameters;
+  if(is_java_implicitly_generic_class_type(type))
+  {
+    const java_implicitly_generic_class_typet &implicitly_generic_class =
+      to_java_implicitly_generic_class_type(to_java_class_type(type));
+    generic_parameters.insert(
+      generic_parameters.end(),
+      implicitly_generic_class.implicit_generic_types().begin(),
+      implicitly_generic_class.implicit_generic_types().end());
+  }
+  if(is_java_generic_class_type(type))
+  {
+    const java_generic_class_typet &generic_class =
+      to_java_generic_class_type(to_java_class_type(type));
+    generic_parameters.insert(
+      generic_parameters.end(),
+      generic_class.generic_types().begin(),
+      generic_class.generic_types().end());
+  }
+  return generic_parameters;
+}
+
+/// Add pairs to the controlled map. Own the keys and pop from their stack
+/// on destruction; otherwise do nothing.
+/// \param parameters generic parameters that are the keys of the pairs to add
+/// \param types a type to add for each parameter
+const void generic_parameter_specialization_map_keyst::insert_pairs(
+  const std::vector<java_generic_parametert> &parameters,
+  const std::vector<reference_typet> &types)
+{
+  INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+  PRECONDITION(parameters.size() == types.size());
+
+  // Pair up the parameters and types for easier manipulation later
+  std::vector<std::pair<java_generic_parametert, reference_typet>> pairs;
+  pairs.reserve(parameters.size());
+  std::transform(
+    parameters.begin(),
+    parameters.end(),
+    types.begin(),
+    std::back_inserter(pairs),
+    [&](java_generic_parametert param, reference_typet type)
+    {
+      return std::make_pair(param, type);
+    });
+
+  for(const auto &pair : pairs)
+  {
+    // Only add the pair if the type is not the parameter itself, e.g.,
+    // pair.first = pair.second = java::A::T. This can happen for example
+    // when initiating a pointer to an implicitly java generic class type
+    // in gen_nondet_init and would result in a loop when the map is used
+    // to look up the type of the parameter.
+    if(
+      !(is_java_generic_parameter(pair.second) &&
+        to_java_generic_parameter(pair.second).get_name() ==
+          pair.first.get_name()))
+    {
+      const irep_idt &key = pair.first.get_name();
+      if(generic_parameter_specialization_map.count(key) == 0)
+        generic_parameter_specialization_map.emplace(
+          key, std::stack<reference_typet>());
+      (*generic_parameter_specialization_map.find(key))
+        .second.push(pair.second);
+
+      // We added something, so pop it when this is destroyed:
+      erase_keys.push_back(key);
+    }
+  }
+}
+
+/// Add pairs of parameters and their types for a generic pointer type to the
```

Comment:
Perhaps improve this comment by explaining that the pairs are for each generic parameter